File: services/websocket/websocket_client.js
/**
 * WebSocket Client for MT4 Market Data
 * Browser and Node.js compatible
 */

class MT4WebSocketClient {
    constructor(url = 'ws://localhost:8765', options = {}) {
        this.url = url;
        this.options = {
            reconnectInterval: 5000,
            maxReconnectAttempts: 10,
            heartbeatInterval: 30000,
            debug: false,
            ...options
        };
        
        this.ws = null;
        this.clientId = null;
        this.authenticated = false;
        this.tier = 'free';
        this.subscriptions = new Set();
        this.eventHandlers = new Map();
        this.reconnectAttempts = 0;
        this.heartbeatTimer = null;
        this.reconnectTimer = null;
        
        // Default event handlers
        this.on('welcome', (data) => {
            this.clientId = data.client_id;
            this.startHeartbeat();
            if (this.options.debug) {
                console.log('Connected to server:', data);
            }
        });
        
        this.on('auth_success', (data) => {
            this.authenticated = true;
            this.tier = data.tier;
            if (this.options.debug) {
                console.log('Authenticated with tier:', data.tier);
            }
        });
        
        this.on('tick', (data) => {
            this.emit('marketData', data.data);
        });
        
        this.on('error', (data) => {
            console.error('Server error:', data.error);
        });
        
        this.on('rate_limit', (data) => {
            console.warn('Rate limited. Retry after:', data.retry_after, 'seconds');
        });
    }
    
    /**
     * Connect to WebSocket server
     */
    connect() {
        return new Promise((resolve, reject) => {
            try {
                // Use appropriate WebSocket implementation
                const WebSocketImpl = typeof WebSocket !== 'undefined' ? WebSocket : require('ws');
                this.ws = new WebSocketImpl(this.url);
                
                this.ws.onopen = () => {
                    this.reconnectAttempts = 0;
                    resolve();
                };
                
                this.ws.onmessage = (event) => {
                    try {
                        const data = JSON.parse(event.data);
                        this.handleMessage(data);
                    } catch (e) {
                        console.error('Failed to parse message:', e);
                    }
                };
                
                this.ws.onclose = () => {
                    this.stopHeartbeat();
                    this.emit('disconnected');
                    this.handleReconnect();
                };
                
                this.ws.onerror = (error) => {
                    console.error('WebSocket error:', error);
                    reject(error);
                };
                
            } catch (error) {
                reject(error);
            }
        });
    }
    
    /**
     * Disconnect from server
     */
    disconnect() {
        this.stopHeartbeat();
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }
        
        if (this.ws && this.ws.readyState === this.ws.OPEN) {
            this.ws.close();
        }
        
        this.ws = null;
        this.authenticated = false;
        this.subscriptions.clear();
    }
    
    /**
     * Authenticate with JWT token
     */
    authenticate(token) {
        return this.send({
            type: 'authenticate',
            token: token
        });
    }
    
    /**
     * Subscribe to market symbols
     */
    subscribe(symbols) {
        if (!Array.isArray(symbols)) {
            symbols = [symbols];
        }
        
        symbols.forEach(symbol => this.subscriptions.add(symbol.toUpperCase()));
        
        return this.send({
            type: 'subscribe',
            symbols: symbols
        });
    }
    
    /**
     * Unsubscribe from symbols
     */
    unsubscribe(symbols) {
        if (!Array.isArray(symbols)) {
            symbols = [symbols];
        }
        
        symbols.forEach(symbol => this.subscriptions.delete(symbol.toUpperCase()));
        
        return this.send({
            type: 'unsubscribe',
            symbols: symbols
        });
    }
    
    /**
     * Get server statistics (premium feature)
     */
    getStats() {
        return this.send({
            type: 'get_stats'
        });
    }
    
    /**
     * Send message to server
     */
    send(data) {
        return new Promise((resolve, reject) => {
            if (!this.ws || this.ws.readyState !== this.ws.OPEN) {
                reject(new Error('WebSocket not connected'));
                return;
            }
            
            try {
                this.ws.send(JSON.stringify(data));
                resolve();
            } catch (error) {
                reject(error);
            }
        });
    }
    
    /**
     * Register event handler
     */
    on(event, handler) {
        if (!this.eventHandlers.has(event)) {
            this.eventHandlers.set(event, []);
        }
        this.eventHandlers.get(event).push(handler);
    }
    
    /**
     * Remove event handler
     */
    off(event, handler) {
        if (!this.eventHandlers.has(event)) return;
        
        const handlers = this.eventHandlers.get(event);
        const index = handlers.indexOf(handler);
        if (index !== -1) {
            handlers.splice(index, 1);
        }
    }
    
    /**
     * Emit event
     */
    emit(event, data) {
        if (!this.eventHandlers.has(event)) return;
        
        this.eventHandlers.get(event).forEach(handler => {
            try {
                handler(data);
            } catch (error) {
                console.error(`Error in ${event} handler:`, error);
            }
        });
    }
    
    /**
     * Handle incoming message
     */
    handleMessage(data) {
        const messageType = data.type;
        
        if (this.options.debug) {
            console.log('Received:', data);
        }
        
        // Emit specific event
        this.emit(messageType, data);
        
        // Emit generic message event
        this.emit('message', data);
    }
    
    /**
     * Start heartbeat
     */
    startHeartbeat() {
        this.stopHeartbeat();
        
        this.heartbeatTimer = setInterval(() => {
            this.send({ type: 'ping' }).catch(() => {
                // Connection lost
                this.stopHeartbeat();
            });
        }, this.options.heartbeatInterval);
    }
    
    /**
     * Stop heartbeat
     */
    stopHeartbeat() {
        if (this.heartbeatTimer) {
            clearInterval(this.heartbeatTimer);
            this.heartbeatTimer = null;
        }
    }
    
    /**
     * Handle reconnection
     */
    handleReconnect() {
        if (this.reconnectAttempts >= this.options.maxReconnectAttempts) {
            console.error('Max reconnection attempts reached');
            this.emit('maxReconnectAttemptsReached');
            return;
        }
        
        this.reconnectAttempts++;
        const delay = this.options.reconnectInterval * this.reconnectAttempts;
        
        console.log(`Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})`);
        
        this.reconnectTimer = setTimeout(() => {
            this.connect()
                .then(() => {
                    console.log('Reconnected successfully');
                    this.emit('reconnected');
                    
                    // Re-subscribe to symbols
                    if (this.subscriptions.size > 0) {
                        this.subscribe(Array.from(this.subscriptions));
                    }
                })
                .catch(() => {
                    // Will trigger another reconnect attempt
                });
        }, delay);
    }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = MT4WebSocketClient;
}

// Example usage
if (typeof window === 'undefined') {
    // Node.js example
    const client = new MT4WebSocketClient('ws://localhost:8765', { debug: true });
    
    // Register event handlers
    client.on('marketData', (data) => {
        console.log('Market data:', data);
    });
    
    client.on('subscribed', (data) => {
        console.log('Subscribed to:', data.symbols);
    });
    
    // Connect and subscribe
    async function example() {
        try {
            await client.connect();
            
            // Authenticate (optional)
            // await client.authenticate('your-jwt-token');
            
            // Subscribe to symbols
            await client.subscribe(['EURUSD', 'GBPUSD', 'USDJPY']);
            
            // Get stats (premium feature)
            // await client.getStats();
            
        } catch (error) {
            console.error('Connection error:', error);
        }
    }
    
    // Run example
    // example();
}

File: services/websocket/websocket_client.py
#!/usr/bin/env python3
"""
WebSocket Client for MT4 Market Data
Python implementation with asyncio support
"""

import asyncio
import websockets
import json
import time
import jwt
import logging
from typing import Optional, List, Dict, Callable, Set
from datetime import datetime
import aiohttp


class MT4WebSocketClient:
    """Async WebSocket client for MT4 market data"""
    
    def __init__(self, 
                 url: str = "ws://localhost:8765",
                 token: Optional[str] = None,
                 debug: bool = False):
        """
        Initialize WebSocket client
        
        Args:
            url: WebSocket server URL
            token: JWT authentication token
            debug: Enable debug logging
        """
        self.url = url
        self.token = token
        self.debug = debug
        
        self.ws: Optional[websockets.WebSocketClientProtocol] = None
        self.client_id: Optional[str] = None
        self.authenticated = False
        self.tier = 'free'
        self.subscriptions: Set[str] = set()
        
        # Event handlers
        self.handlers: Dict[str, List[Callable]] = {}
        
        # Connection settings
        self.reconnect_interval = 5
        self.max_reconnect_attempts = 10
        self.heartbeat_interval = 30
        
        # Tasks
        self.heartbeat_task: Optional[asyncio.Task] = None
        self.receive_task: Optional[asyncio.Task] = None
        
        # Setup logging
        self.logger = logging.getLogger(__name__)
        if debug:
            self.logger.setLevel(logging.DEBUG)
    
    def on(self, event: str, handler: Callable):
        """Register event handler"""
        if event not in self.handlers:
            self.handlers[event] = []
        self.handlers[event].append(handler)
    
    def off(self, event: str, handler: Callable):
        """Remove event handler"""
        if event in self.handlers and handler in self.handlers[event]:
            self.handlers[event].remove(handler)
    
    async def emit(self, event: str, data: Any):
        """Emit event to handlers"""
        if event in self.handlers:
            for handler in self.handlers[event]:
                try:
                    if asyncio.iscoroutinefunction(handler):
                        await handler(data)
                    else:
                        handler(data)
                except Exception as e:
                    self.logger.error(f"Error in {event} handler: {e}")
    
    async def connect(self):
        """Connect to WebSocket server"""
        for attempt in range(self.max_reconnect_attempts):
            try:
                self.logger.info(f"Connecting to {self.url}...")
                self.ws = await websockets.connect(self.url)
                
                # Start receive loop
                self.receive_task = asyncio.create_task(self._receive_loop())
                
                # Wait for welcome message
                await asyncio.sleep(0.1)
                
                # Authenticate if token provided
                if self.token and not self.authenticated:
                    await self.authenticate(self.token)
                
                # Re-subscribe to previous subscriptions
                if self.subscriptions:
                    await self.subscribe(list(self.subscriptions))
                
                self.logger.info("Connected successfully")
                await self.emit('connected', None)
                
                return
                
            except Exception as e:
                self.logger.error(f"Connection failed (attempt {attempt + 1}): {e}")
                if attempt < self.max_reconnect_attempts - 1:
                    await asyncio.sleep(self.reconnect_interval * (attempt + 1))
                else:
                    raise ConnectionError(f"Failed to connect after {self.max_reconnect_attempts} attempts")
    
    async def disconnect(self):
        """Disconnect from server"""
        if self.heartbeat_task:
            self.heartbeat_task.cancel()
        
        if self.receive_task:
            self.receive_task.cancel()
        
        if self.ws:
            await self.ws.close()
            self.ws = None
        
        self.authenticated = False
        self.client_id = None
        
        await self.emit('disconnected', None)
    
    async def authenticate(self, token: Optional[str] = None):
        """Authenticate with JWT token"""
        if token:
            self.token = token
        
        if not self.token:
            raise ValueError("No token provided")
        
        await self.send({
            'type': 'authenticate',
            'token': self.token
        })
    
    async def subscribe(self, symbols: List[str]):
        """Subscribe to market symbols"""
        if isinstance(symbols, str):
            symbols = [symbols]
        
        symbols = [s.upper() for s in symbols]
        self.subscriptions.update(symbols)
        
        await self.send({
            'type': 'subscribe',
            'symbols': symbols
        })
    
    async def unsubscribe(self, symbols: List[str]):
        """Unsubscribe from symbols"""
        if isinstance(symbols, str):
            symbols = [symbols]
        
        symbols = [s.upper() for s in symbols]
        for symbol in symbols:
            self.subscriptions.discard(symbol)
        
        await self.send({
            'type': 'unsubscribe',
            'symbols': symbols
        })
    
    async def get_stats(self):
        """Request server statistics (premium feature)"""
        await self.send({
            'type': 'get_stats'
        })
    
    async def send(self, data: Dict):
        """Send message to server"""
        if not self.ws:
            raise ConnectionError("Not connected")
        
        if self.debug:
            self.logger.debug(f"Sending: {data}")
        
        await self.ws.send(json.dumps(data))
    
    async def _receive_loop(self):
        """Receive messages from server"""
        try:
            async for message in self.ws:
                try:
                    data = json.loads(message)
                    await self._handle_message(data)
                except json.JSONDecodeError as e:
                    self.logger.error(f"Failed to parse message: {e}")
                except Exception as e:
                    self.logger.error(f"Error handling message: {e}")
        
        except websockets.exceptions.ConnectionClosed:
            self.logger.warning("Connection closed")
            await self.emit('disconnected', None)
            # Attempt reconnection
            asyncio.create_task(self.connect())
    
    async def _handle_message(self, data: Dict):
        """Handle incoming message"""
        if self.debug:
            self.logger.debug(f"Received: {data}")
        
        msg_type = data.get('type')
        
        # Handle specific message types
        if msg_type == 'welcome':
            self.client_id = data.get('client_id')
            self.heartbeat_interval = data.get('heartbeat_interval', 30)
            # Start heartbeat
            if self.heartbeat_task:
                self.heartbeat_task.cancel()
            self.heartbeat_task = asyncio.create_task(self._heartbeat_loop())
        
        elif msg_type == 'auth_success':
            self.authenticated = True
            self.tier = data.get('tier', 'free')
            self.logger.info(f"Authenticated with tier: {self.tier}")
        
        elif msg_type == 'auth_failed':
            self.authenticated = False
            self.logger.error(f"Authentication failed: {data.get('error')}")
        
        elif msg_type == 'tick':
            # Emit market data event
            await self.emit('market_data', data.get('data'))
        
        elif msg_type == 'subscribed':
            self.logger.info(f"Subscribed to: {data.get('symbols')}")
        
        elif msg_type == 'error':
            self.logger.error(f"Server error: {data.get('error')}")
        
        elif msg_type == 'rate_limit':
            self.logger.warning(f"Rate limited. Retry after: {data.get('retry_after')}s")
        
        # Emit generic message event
        await self.emit('message', data)
        
        # Emit specific event
        if msg_type:
            await self.emit(msg_type, data)
    
    async def _heartbeat_loop(self):
        """Send periodic heartbeat"""
        while True:
            try:
                await asyncio.sleep(self.heartbeat_interval)
                await self.send({'type': 'ping'})
            except Exception as e:
                self.logger.error(f"Heartbeat error: {e}")
                break


class MT4WebSocketSync:
    """Synchronous wrapper for WebSocket client"""
    
    def __init__(self, url: str = "ws://localhost:8765", token: Optional[str] = None):
        self.client = MT4WebSocketClient(url, token)
        self.loop = asyncio.new_event_loop()
        self.thread = None
    
    def connect(self):
        """Connect synchronously"""
        self.loop.run_until_complete(self.client.connect())
    
    def disconnect(self):
        """Disconnect synchronously"""
        self.loop.run_until_complete(self.client.disconnect())
    
    def subscribe(self, symbols: List[str]):
        """Subscribe synchronously"""
        self.loop.run_until_complete(self.client.subscribe(symbols))
    
    def on_market_data(self, handler: Callable):
        """Register market data handler"""
        self.client.on('market_data', handler)


def generate_test_token(user_id: str = "test_user", tier: str = "free") -> str:
    """Generate test JWT token"""
    payload = {
        'user_id': user_id,
        'tier': tier,
        'exp': datetime.utcnow().timestamp() + 3600,
        'iat': datetime.utcnow().timestamp()
    }
    
    return jwt.encode(payload, 'your-secret-key', algorithm='HS256')


# Example usage
async def example():
    """Example WebSocket client usage"""
    # Create client
    client = MT4WebSocketClient("ws://localhost:8765", debug=True)
    
    # Register handlers
    @client.on('market_data')
    async def handle_market_data(data):
        print(f"Market data: {data}")
    
    @client.on('stats')
    async def handle_stats(data):
        print(f"Stats: {data}")
    
    try:
        # Connect
        await client.connect()
        
        # Generate and use test token
        token = generate_test_token("test_user", "premium")
        await client.authenticate(token)
        
        # Subscribe to symbols
        await client.subscribe(['EURUSD', 'GBPUSD', 'USDJPY'])
        
        # Get stats (premium feature)
        await client.get_stats()
        
        # Keep running
        await asyncio.sleep(60)
        
    except Exception as e:
        print(f"Error: {e}")
    
    finally:
        await client.disconnect()


# REST API client for comparison
class MT4RestClient:
    """REST API client with WebSocket fallback"""
    
    def __init__(self, base_url: str = "http://localhost:5000", api_key: Optional[str] = None):
        self.base_url = base_url
        self.api_key = api_key
        self.session = aiohttp.ClientSession()
        
        if api_key:
            self.session.headers.update({'X-API-Key': api_key})
    
    async def get_tick(self, symbol: str) -> Dict:
        """Get latest tick via REST API"""
        async with self.session.get(f"{self.base_url}/api/v1/market/tick/{symbol}") as resp:
            return await resp.json()
    
    async def get_ticks(self, symbols: List[str]) -> Dict:
        """Get multiple ticks"""
        params = {'symbols': ','.join(symbols)}
        async with self.session.get(f"{self.base_url}/api/v1/market/ticks", params=params) as resp:
            return await resp.json()
    
    async def close(self):
        """Close session"""
        await self.session.close()


if __name__ == "__main__":
    # Run example
    logging.basicConfig(level=logging.INFO)
    asyncio.run(example())